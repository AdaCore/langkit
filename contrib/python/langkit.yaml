lkt_spec:
  entry_point: nodes.lkt
  source_dirs: [.]
library:
  language_name: Python
warnings:
  undocumented-nodes: false
