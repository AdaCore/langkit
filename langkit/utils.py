"""
This file contains general utility functions. It is basically the support
library for langkit, aggregating general python utilities that we could not
find in the standard library.
"""

from __future__ import (absolute_import, division, print_function,
                        unicode_literals)

from copy import copy
from itertools import takewhile

import inspect
import sys


class StructEq(object):
    """
    Mixin for structural equality.
    """
    def __eq__(self, other):
        if type(other) is type(self):
            if hasattr(self, "_eq_keys"):
                eq_keys = self._eq_keys
            elif hasattr(self, "_excl_eq_keys"):
                eq_keys = set(self.__dict__.keys()) ^ set(self._excl_eq_keys)
            else:
                return self.__dict__ == other.__dict__

            return all(v == other.__dict__[k] for k, v in self.__dict__
                       if k in eq_keys)

        return False


def unescape(char):
    """
    Unescape char if it is escaped.

    :param str char: An eventually escaped character.
    :rtype: str
    """
    if char[0] == "\\":
        return char[1:]
    return char


def copy_with(obj, **kwargs):
    """
    Copy an object, and add every key value association passed in kwargs to it.

    :param dict[str, Any] kwargs: key value associations to attach to the
        object.
    :param T obj: The object to copy.
    :rtype: T

    Example use::

        class A(object):
            def __init__(self, a, b):
                self.a = a
                self.b = b

        inst = A(12, 15)
        inst_2 = copy_with(inst, a=16)
        print inst_2.a
        # Prints "16"
        print inst_2.b
        # Prints "12"
    """
    c = copy(obj)

    for k, v in kwargs.items():
        setattr(c, k, v)

    return c


class Colors(object):
    """
    Utility escape sequences to color output in terminal.
    """
    ENDC = '\033[0m'
    BOLD = '\033[1m'

    RED = '\033[91m'
    GREEN = '\033[92m'
    YELLOW = '\033[93m'
    BLUE = '\033[94m'
    MAGENTA = '\033[95m'
    CYAN = '\033[96m'
    GREY = '\033[97m'

    HEADER = MAGENTA
    OKBLUE = BLUE
    OKGREEN = GREEN
    WARNING = YELLOW
    FAIL = RED

    @classmethod
    def disable_colors(cls):
        """
        Reset color codes to empty strings.
        """
        for name in dir(cls):
            value = getattr(cls, name)
            if (all(c.isalpha() for c in name) and
                    name.upper() == name and
                    isinstance(value, basestring)):
                setattr(cls, name, '')


if not sys.stdout.isatty() or not sys.stderr.isatty():
    Colors.disable_colors()


def col(msg, color):
    """
    Utility function that return a string colored with the proper escape
    sequences, for VT100 compatible terminals.

    :param str msg: The message to print.
    :param color: An escape sequence corresponding to the proper color. Pick
        one in the Colors class.
    :rtype: str
    """
    return "{0}{1}{2}".format(color, msg, Colors.ENDC)


def printcol(msg, color):
    """
    Utility print function that will print `msg` in color `color`.
    :param basestring msg: The message to print.
    :param basestring color: The color escape sequence from the enum class
        Colors which represents the color to use.
    :return: The color-escaped string, resetting the color to blank at the end.
    :rtype: basestring
    """
    print(col(msg, color))

    # Colored messages are used to show the user how the compilation process is
    # going, so flushing on a regular basis matters. When \n-based flushing is
    # disabled (because of a pipe, for instance), force flushing here.
    sys.stdout.flush()


def common_ancestor(*classes):
    """
    Return the bottom-most common parent class for all `classes`.

    Note that this considers only the first parent for each class, as if there
    was only single inheritance.

    :param classes: The classes for which we are searching a common ancestor.
    :type classes: list[types.ClassType]
    :rtype: types.ClassType
    """
    def rmro(klass):
        single_mro = []
        cls = klass
        while True:
            single_mro.append(cls)
            try:
                cls = cls.__bases__[0]
            except IndexError:
                break
        return reversed(single_mro)

    result = list(takewhile(lambda a: len(set(a)) == 1,
                            zip(*map(rmro, classes))))[-1][0]
    return result


def memoized(func):
    """
    Decorator to memoize a function.
    This function must be passed only hashable arguments.

    :param func: The function to decorate.
    """
    cache = {}

    if not getattr(memoized, "caches", None):
        memoized.caches = []

    memoized.caches.append(cache)

    def wrapper(*args, **kwargs):
        key = (args, tuple(kwargs.items()))
        try:
            result = cache[key]
        except KeyError:
            result = func(*args, **kwargs)
            cache[key] = result
            result = result
        return result

    wrapper.reset_cache = lambda: cache.clear()

    return wrapper


def reset_memoized():
    if getattr(memoized, "caches", None):
        for cache in memoized.caches:
            cache.clear()


def type_check_exact(klass):
    """
    Return a predicate that will return true if its parameter is exactly egal
    to `klass`.

    :param type klass: Class to check against.
    :rtype: (T) -> bool
    """
    return lambda t: t and t == klass


def type_check(klass):
    """
    Return a predicate that will return true if its parameter is a subclass
    of `klass`.
    :param type klass: Class to check against.
    :rtype: (T) -> bool
    """
    return lambda t: t and issubclass(t, klass)


def type_check_instance(klass):
    """
    Return a predicate that will return true if its parameter is a subclass
    of `klass`.
    :param type klass: Class to check against.
    :rtype: (T) -> bool
    """
    return lambda t: isinstance(t, klass)


def dispatch_on_type(type, type_to_action_assocs, exception=None):
    """
    Dispatch on the type parameter, execute the corresponding action
    depending on the type. Every type in type_to_action_assocs will be
    tested in turn. If type is a subtype of one of them, then the
    corresponding action will be executed.

    :param type: The type to dispatch upon.

    :param type_to_action_assocs: An association of types to actions that
        returns something.
    :type type_to_action_assocs: list[(type, (type) -> T)]

    :param Exception exception: The exception to raise in case the type is
        not in the dispatch table.

    :rtype: T
    """
    exception = exception or Exception(
        "Error in dispatch_on_type: {} not handled".format(type)
    )
    for target_type, action in type_to_action_assocs:
        if issubclass(type, target_type):
            return action(type)
    raise exception


def assert_type(obj, typ):
    """
    Ensure that obj is a subclass or instance of type, depending on whether obj
    is a type or an instance. Return obj. This allows a few things compared to
    regular asserts:

    1. The type of the return is correctly inferred by Pycharm.
    2. You can use a "fluent interface" style for style checks, using::

        b = assert_type(c, int) + 15

    rather than::

        assert isinstance(c, int)
        b = c + 15

    3. Provides a relatively informative error message by default.

    :param Any obj: The object to check.
    :param T typ: Type parameter. The expected type of obj.
    :rtype: T
    """
    if issubclass(type(obj), type):
        assert issubclass(obj, typ), (
            "Wrong base type for type {}, expected {}".format(obj, typ)
        )
    else:
        assert isinstance(obj, typ), (
            "Wrong type for object {}, expected {}, found {}".format(
                obj, typ, type(obj)
            )
        )
    return obj


class TypeSet(object):
    """
    This class is an helper for when you need to check wether an abstract
    operation was fulfilled for a whole hierarchy of classes, given that if
    the operation is fulfilled for all subclasses, then we consider that it is
    done for the parent class.

    This allows to automate the logic of, for example, checking if abstract
    properties are overridden properly, or if a match expression handles all
    cases.
    """

    def __init__(self):
        # Working set of ASTNode subclasses for the types that are covered by
        # matchers. Updated as we go through the list of matchers.
        self.matched_types = set()

    def include(self, t):
        """
        Include a class and all of its subclasses.

        If t is the last subclass for some base class, this adds the parent
        subclass. This makes sense as once all concrete subclasses of the
        abstract A type are handled, it is true that A is handled.

        Return whether t was already present in self.

        :param ASTNode|None t: Type parameter.
        :rtype: bool
        """
        if t in self.matched_types:
            return True

        # Include "t" and all its subclasses
        self.matched_types.add(t)
        for child in t.subclasses:
            self.include(child)

        # Include its parents if all their children are matched
        parents = list(t.get_inheritance_chain())[:-1]
        for parent in reversed(parents):
            if parent in self.matched_types:
                break
            subclasses = set(parent.subclasses)
            if not subclasses.issubset(self.matched_types):
                break
            # If we reach this point, all parent's subclasses are matched,
            # so we can state that parent itself is always matched.
            self.matched_types.add(parent)

        return False

    def unmatched_types(self, t):
        """
        Return the set of t subclasses that are not matched by any matcher.

        Omit subclasses when none of them are matched: only the parent is
        returned in this case, so that we don't flood users with whole
        hierarchies of classes.

        :param ASTNode t: Type parameter.
        :rtype: set[ASTNode]
        """
        if t in self.matched_types:
            return set()

        subclasses = set(t.subclasses)
        if subclasses & self.matched_types:
            result = set()
            for cls in subclasses:
                result.update(self.unmatched_types(cls))
        else:
            result = {t}
        return result


def issubtype(type, parent_type):
    """
    Helper wrapper around issubclass, because issubclass will raise an
    exception if compared object is not a class, which is extremely tedious for
    us in langkit.

    :param type: The object or type to check against parent_type.
    :param type parent_type: The supposed parent_type of type.
    """
    return inspect.isclass(type) and issubclass(type, parent_type)


class DictProxy(object):
    """
    Util class to be able to access dict keys via the attribute notation in
    Python. Used in the property DSL.
    """
    def __init__(self, dct):
        self.dct = dct

    def __getattr__(self, attr):
        return self.dct[attr]


def not_implemented_error(self_or_cls, method):
    """
    Return a NotImplementedError instance to explain that `self or_cls` must
    override method `method`.

    :param self or_cls: Class that does not implement `method`, or the instance
        of such a class.
    :param method: Method object for the method that should be overriden.
    :rtype: NotImplementedError
    """
    cls = self_or_cls if inspect.isclass(self_or_cls) else type(self_or_cls)
    return NotImplementedError('{} must override method {}'.format(
        cls.__name__, method.__name__
    ))
