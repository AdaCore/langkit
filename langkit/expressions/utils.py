from __future__ import annotations

import langkit.expressions as E


def inc_ref(var: E.VariableExpr | E.LocalVars.LocalVar) -> str:
    """
    Shortcut for expression rendering.

    If ``var`` is a ref-counted variable, return Ada code to inc-ref it
    (one statement including trailing semi-colon). Otherwise, return an empty
    string.
    """
    var_name = (
        var.codegen_name if isinstance(var, E.LocalVars.LocalVar) else var.name
    )
    return f"Inc_Ref ({var_name});" if var.type.is_refcounted else ""


def assign_var(
    var: E.VariableExpr | E.LocalVars.LocalVar,
    expr: str,
    requires_incref: bool = True,
) -> str:
    """
    Shortcut for expression rendering.

    Return Ada code to assign the given `expr` to the given `var`.  If
    `requires_incref` is True and `var` is a ref-counted variable, also emit
    code to inc-ref it. The result is a sequence of Ada statements, including
    trailing semi-colon.
    """
    var_name = (
        var.codegen_name if isinstance(var, E.LocalVars.LocalVar) else var.name
    )
    return f"{var_name} := {expr}; {inc_ref(var) if requires_incref else ''}"


def array_aggr(exprs: list[str]) -> str:
    """
    Shortcut to format an array aggregate expression.
    """
    return (
        "({})".format(", ".join(["{}"] * len(exprs)))
        if exprs
        else "(1 .. 0 => <>)"
    )
