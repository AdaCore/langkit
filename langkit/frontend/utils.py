from __future__ import annotations

from collections import OrderedDict
import os.path

from langkit.compile_context import CompileCtx
from langkit.config import LktSpecConfig
from langkit.diagnostics import (
    Location,
    error,
    errors_checkpoint,
    non_blocking_error,
)
import langkit.names as names

import liblktlang as L


def name_from_lower(ctx: CompileCtx, kind: str, id: L.Id) -> names.Name:
    """
    Validate "id" as a lower-case name and return the corresponding ``Name``
    instance.
    """
    try:
        names.check_common(id.text)
    except ValueError as exc:
        error(str(exc), location=id)
    try:
        names.check_lower(id.text)
    except ValueError:
        error(f"lower case expected for {kind} names", location=id)
    return names.Name.from_lower(id.text)


def name_from_camel(ctx: CompileCtx, kind: str, id: L.Id) -> names.Name:
    """
    Validate "id" as a camel-case name and return the corresponding ``Name``
    instance.
    """
    try:
        names.check_common(id.text)
    except ValueError as exc:
        error(str(exc), location=id)
    try:
        names.check_camel(id.text)
    except ValueError as exc:
        error(f"camel case expected for {kind} names ({exc})", location=id)
    return names.Name.from_camel(id.text)


def arg_name_from_expr(e: L.Expr) -> L.Id:
    """
    Return the argument name for a given argument value expression.

    This assumes that ``e`` is the ``f_value`` argument of a named ``Argument``
    node, and return the ``f_name`` field for that argument.
    """
    arg = e.parent
    assert isinstance(arg, L.Argument)
    assert arg.f_name is not None
    return arg.f_name


def lkt_doc(decl: L.Decl) -> str:
    """
    Return the documentation attached to the ``decl`` node. This is an empty
    string if the docstring is missing.

    :param decl: Declaration to process.
    """
    from langkit.frontend.static import denoted_str

    full_decl = decl.parent
    assert isinstance(full_decl, L.FullDecl)
    return "" if full_decl.f_doc is None else denoted_str(full_decl.f_doc)


def load_lkt(config: LktSpecConfig) -> list[L.AnalysisUnit]:
    """
    Load a Lktlang source file and return the closure of Lkt units referenced.
    Raise a DiagnosticError if there are parsing errors.

    :param lkt_file: Name of the file to parse.
    """
    units_map = OrderedDict()
    diagnostics = []

    def process_unit(unit: L.AnalysisUnit) -> None:
        if unit.filename in units_map:
            return

        # Register this unit and its diagnostics
        units_map[unit.filename] = unit
        for d in unit.diagnostics:
            diagnostics.append((unit, d))

        # Recursively process the units it imports. In case of parsing error,
        # just stop the recursion: the collection of diagnostics is enough.
        if not unit.diagnostics:
            assert isinstance(unit.root, L.LangkitRoot)
            import_stmts = list(unit.root.f_imports)
            for imp in import_stmts:
                process_unit(imp.p_referenced_unit)

    # Give Liblktlang access to the Lkt files to analyze
    old_path = os.environ.get("LKT_PATH", "")
    os.environ["LKT_PATH"] = os.path.pathsep.join(config.source_dirs)

    # Load ``lkt_file`` and all the units it references, transitively
    ctx = L.AnalysisContext(unit_provider=L.UnitProvider.from_lkt_path())
    process_unit(ctx.get_from_file(config.entry_point))

    os.environ["LKT_PATH"] = old_path

    # Forward potential lexing/parsing errors to our diagnostics system
    for u, d in diagnostics:
        non_blocking_error(
            d.message, Location.from_sloc_range(u, d.sloc_range)
        )
    errors_checkpoint()
    return list(units_map.values())
