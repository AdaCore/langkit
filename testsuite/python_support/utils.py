from __future__ import absolute_import, division, print_function

import os
import os.path
import shutil
import subprocess
import sys

import langkit
import langkit.compile_context
from langkit.compile_context import CompileCtx
from langkit.diagnostics import DiagnosticError, WarningSet
from langkit.libmanage import ManageScript

from testsuite_support.valgrind import valgrind_cmd


default_warning_set = WarningSet()

# We don't want to be forced to provide dummy docs for nodes and public
# properties in testcases.
default_warning_set.disable(WarningSet.undocumented_nodes)
default_warning_set.disable(WarningSet.undocumented_public_properties)

pretty_print = bool(int(os.environ.get('LANGKIT_PRETTY_PRINT', '0')))

project_template = """
with "libfoolang";

project Gen is
    for Languages use ("Ada");
    for Source_Dirs use (".");
    for Object_Dir use "obj";
    for Main use ({main_sources});

    package Compiler is
        for Default_Switches ("Ada") use
          ("-g", "-O0", "-gnata", "-gnatwae", "-gnatyg");
    end Compiler;
end Gen;
"""


# Determine where to find the root directory for Langkit sources
langkit_root = os.environ.get('LANGKIT_ROOT_DIR')
if not langkit_root:
    test_dir = os.path.dirname(os.path.abspath(__file__))
    testsuite_dir = os.path.dirname(test_dir)
    langkit_root = os.path.dirname(testsuite_dir)


def prepare_context(grammar, lexer=None, warning_set=default_warning_set,
                    symbol_canonicalizer=None):
    """
    Create a compile context and prepare the build directory for code
    generation.

    :param langkit.parsers.Grammar grammar: The language grammar to use for
        this context.

    :param langkit.lexer.Lexer lexer: The language lexer to use for this
        context.

    :param WarningSet warning_set: Set of warnings to emit.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    # Have a clean build directory
    if os.path.exists('build'):
        shutil.rmtree('build')
    os.mkdir('build')

    # Try to emit code
    ctx = CompileCtx(lang_name='Foo', lexer=lexer, grammar=grammar,
                     symbol_canonicalizer=symbol_canonicalizer)
    ctx.warnings = warning_set
    ctx.pretty_print = pretty_print

    return ctx


def emit_and_print_errors(grammar, lexer=None,
                          warning_set=default_warning_set,
                          generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for CTX. Return whether this was successful.

    :param langkit.parsers.Grammar grammar_fn: The language grammar to use.

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        Use `lexer_example.foo_lexer` if left to None.

    :param WarningSet warning_set: Set of warnings to emit.

    :param bool generate_unparser: Whether to generate unparser.

    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.

    :rtype: bool
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    try:
        ctx = prepare_context(grammar, lexer, warning_set,
                              symbol_canonicalizer=symbol_canonicalizer)
        ctx.emit('build', generate_unparser=generate_unparser)
        # ... and tell about how it went
    except DiagnosticError:
        # If there is a diagnostic error, don't say anything, the diagnostics
        # are enough.
        return False
    else:
        print('Code generation was successful')
        return True
    finally:
        lexer._dfa_code = None
        langkit.reset()


def build(grammar, lexer=None, warning_set=default_warning_set):
    """
    Shortcut for `build_and_run` to only build.
    """
    build_and_run(grammar, lexer=lexer, warning_set=warning_set)


def build_and_run(grammar, py_script=None, ada_main=None, lexer=None,
                  ocaml_main=None, warning_set=default_warning_set,
                  generate_unparser=False, symbol_canonicalizer=None):
    """
    Compile and emit code for `ctx` and build the generated library. Then,
    execute the provided scripts/programs, if any.

    An exception is raised if any step fails (the script must return code 0).

    :param langkit.lexer.Lexer lexer: The lexer to use along with the grammar.
        See emit_and_print_errors.
    :param None|str py_script: If not None, name of the Python script to run
        with the built library available.
    :param None|str|list[str] ada_main: If not None, list of name of main
        source files for Ada programs to build and run with the generated
        library. If the input is a single string, consider it's a single mail
        source file.
    :param WarningSet warning_set: Set of warnings to emit.
    :param bool generate_unparser: Whether to generate unparser.
    :param langkit.compile_context.LibraryEntity|None symbol_canonicalizer:
        Symbol canoncalizes to use for this context, if any.
    """

    if lexer is None:
        from lexer_example import foo_lexer
        lexer = foo_lexer

    ctx = prepare_context(grammar, lexer, warning_set,
                          symbol_canonicalizer=symbol_canonicalizer)

    class Manage(ManageScript):
        def create_context(self, args):
            return ctx

    m = Manage(override_lang_source_dir=False)

    extensions_dir = os.path.abspath('extensions')
    if os.path.isdir(extensions_dir):
        ctx.extensions_dir = extensions_dir

    # First build the library. Forward all test.py's arguments to the libmanage
    # call so that manual testcase runs can pass "-g", for instance.
    argv = sys.argv[1:] + ['--full-error-traces', '-vnone', 'make']
    for w in WarningSet.available_warnings:
        argv.append('-{}{}'.format('W' if w in warning_set else 'w', w.name))
    if not pretty_print:
        argv.append('--no-pretty-print')
    if generate_unparser:
        argv.append('--generate-unparser')
    m.run(argv)

    # Flush stdout and stderr, so that diagnostics appear deterministically
    # before the script/program output.
    sys.stdout.flush()
    sys.stderr.flush()

    # Write a "setenv" script to make developper investigation convenient
    with open('setenv.sh', 'w') as f:
        m.write_setenv(f)

    env = m.derived_env()

    def run(*argv):
        subprocess.check_call(argv, env=env)

    if py_script is not None:
        # Run the Python script. Note that in order to use the generated
        # library, we have to use the special Python interpreter the testsuite
        # provides us. See the corresponding code in
        # testuite_support/python_driver.py.
        python_interpreter = os.environ['PYTHON_INTERPRETER']
        run(python_interpreter, py_script)

    if ada_main is not None:
        if isinstance(ada_main, str):
            ada_main = [ada_main]

        # Generate a project file to build the given Ada main and then run
        # the program.
        with open('gen.gpr', 'w') as f:
            f.write(project_template.format(
                main_sources=', '.join('"{}"'.format(m) for m in ada_main)
            ))
        run('gprbuild', '-Pgen', '-q', '-p',
            '-XLIBRARY_TYPE=relocatable', '-XXMLADA_BUILD=relocatable')

        for i, m in enumerate(ada_main):
            assert m.endswith('.adb')
            if i > 0:
                print('')
            if len(ada_main) > 1:
                print('== {} =='.format(m))
            sys.stdout.flush()
            argv = [os.path.join('obj', m[:-4])]
            if os.environ.get('VALGRIND_ENABLED'):
                argv = valgrind_cmd(argv)
            run(*argv)

    if ocaml_main is not None:
        run('dune', 'exec', '--display', 'quiet', '--root', '.',
            './{}.exe'.format(ocaml_main))


def add_gpr_path(dirname):
    """
    Prepend the given directory name to the ``GPR_PROJECT_PATH`` environment
    variable.
    """
    old_path = os.environ.get('GPR_PROJECT_PATH')
    os.environ['GPR_PROJECT_PATH'] = (
        '{}{}{}'.format(dirname, os.path.pathsep, old_path)
        if old_path else dirname)
