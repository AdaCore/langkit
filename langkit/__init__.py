def reset():
    """
    Reset global state in Langkit.

    TODO: this is a hack to workaround another hack. At some point in the
    future, we should get rid of this global state in Langkit.
    """

    import langkit.compile_context
    import langkit.compiled_types
    import langkit.expressions.base
    import langkit.lexer
    import langkit.utils

    langkit.utils.reset_memoized()

    langkit.compiled_types.CompiledTypeRepo.reset()
    langkit.compiled_types.create_builtin_types()

    langkit.expressions.base.Entity.unfreeze()
    langkit.expressions.base.Self.unfreeze()

    langkit.lexer.LexerToken.reset()

    langkit.compile_context.compile_ctx = None
