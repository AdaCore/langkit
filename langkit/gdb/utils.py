from langkit.names import Name
from langkit.utils import Colors, col


def expr_repr(expr):
    """
    Return a colored repr for an expression.

    :type expr: langkit.gdb.state.ExpressionEvaluation
    :rtype: str
    """
    return col(expr.expr_repr, Colors.CYAN)


def name_repr(expr):
    """
    Return a colored repr for a binding name.

    :type expr: langkit.gdb.state.Binding
    :rtype: str
    """
    return col(expr.dsl_name, Colors.GREEN)


def prop_repr(prop):
    """
    Return a colored repr for a property name.

    :type prop: langkit.gdb.debug_info.Property
    :rtype: str
    """
    prefix = '[dispatcher]'
    if prop.name.startswith(prefix):
        name = prop.name[len(prefix):]
        label = 'dispatch property for {}'.format(name)
    else:
        label = prop.name
    return col(col(label, Colors.RED), Colors.BOLD)


def adaify_name(context, name):
    """
    Turn a symbol name like a__b into an Ada-like name such as A.B.
    Also strip the $LIB_NAME.Analysis prefix, if present.

    :type name: str
    :rtype: str
    """
    pfx = context.analysis_prefix
    if name.startswith(pfx):
        name = name[len(pfx):]
    chunks = name.split('__')
    return '.'.join(Name.from_lower(c).camel_with_underscores
                    for c in chunks)


def dereference_fat_array_ptr(fat_ptr):
    """
    Dereference an array reference materialized as a fat pointer.
    """
    lower_bound = int(fat_ptr['P_BOUNDS']['LB0'])
    upper_bound = int(fat_ptr['P_BOUNDS']['UB0'])
    array_ptr = fat_ptr['P_ARRAY']
    element_type = array_ptr.type.target().target()
    array_ptr_type = element_type.array(lower_bound, upper_bound).pointer()
    return array_ptr.cast(array_ptr_type).dereference()
