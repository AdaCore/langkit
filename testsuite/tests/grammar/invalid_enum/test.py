from __future__ import absolute_import, division, print_function

from langkit.dsl import ASTNode

from utils import emit_and_print_errors


for lkt_file in ('no-dot.lkt', 'bad-prefix.lkt', 'bad-alt.lkt'):
    print('== {} =='.format(lkt_file))

    class FooNode(ASTNode):
        pass

    class Example(FooNode):
        enum_node = True
        alternatives = ['example', 'null', 'def']

    emit_and_print_errors(lkt_file=lkt_file)
    print('')


print('Done')
