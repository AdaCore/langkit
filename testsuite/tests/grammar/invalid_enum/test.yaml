driver: lkt_compile
description: |
    Test that invalid references to enum nodes in grammars are properly
    rejected.
