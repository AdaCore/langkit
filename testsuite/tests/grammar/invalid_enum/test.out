== no-dot.lkt ==
no-dot.lkt:4:29: error: Parsers cannot create abstract nodes
3 |     @main_rule main_rule <- Example("example")
  |                             ^^^^^^^^^^^^^^^^^^


== bad-prefix.lkt ==
bad-prefix.lkt:4:29: error: Unknown node: NoSuchNode
3 |     @main_rule main_rule <- NoSuchNode.Example("example")
  |                             ^^^^^^^^^^


== bad-alt.lkt ==
bad-alt.lkt:4:37: error: Unknown enum node alternative: NoSuchAlt
3 |     @main_rule main_rule <- Example.NoSuchAlt("example")
  |                                     ^^^^^^^^^


Done
